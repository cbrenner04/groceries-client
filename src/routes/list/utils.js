import { toast } from 'react-toastify';

import axios from '../../utils/api';
import { formatDueBy } from '../../utils/format';

export function itemName(item, listType) {
  return {
    BookList: `${item.title ? `"${item.title}"` : ''} ${item.author || ''}`,
    GroceryList: `${item.quantity || ''} ${item.product || ''}`,
    MusicList:
      `${item.title ? `"${item.title}"` : ''} ${item.artist || ''}` +
      `${item.artist && item.album ? ' - ' : ''}${item.album || ''}`,
    ToDoList: `${item.task}`,
  }[listType].trim();
}

export function mapIncludedCategories(items) {
  const cats = [''];
  items.forEach((item) => {
    if (!item.category) {
      return;
    }
    const cat = item.category.toLowerCase();
    if (!cats.includes(cat)) {
      cats.push(cat);
    }
  });
  return cats;
}

export function categorizeNotPurchasedItems(items, categories) {
  const obj = {};
  categories.forEach((cat) => {
    obj[cat] = [];
  });
  items.forEach((item) => {
    if (!item.category) {
      obj[''].push(item);
      return;
    }
    const cat = item.category.toLowerCase();
    if (!obj[cat]) {
      obj[cat] = [];
    }
    obj[cat].push(item);
  });
  return obj;
}

function performSort(items, sortAttrs) {
  // return when all items are sorted
  if (sortAttrs.length === 0) {
    return items;
  }
  const sortAttr = sortAttrs.pop();
  const sorted = items.sort((a, b) => {
    // the sort from the server comes back with items with number_in_series: `null` at the end of the list
    // without the next two lines this would put those items at the front of the list
    if (a[sortAttr] === null) {
      return 1;
    }
    if (b[sortAttr] === null) {
      return -1;
    }
    const positiveBranch = a[sortAttr] > b[sortAttr] ? 1 : 0;
    return a[sortAttr] < b[sortAttr] ? -1 : positiveBranch;
  });
  return performSort(sorted, sortAttrs);
}

export function sortItems(listType, items) {
  let sortAttrs = [];
  if (listType === 'BookList') {
    sortAttrs = ['author', 'number_in_series', 'title'];
  } else if (listType === 'GroceryList') {
    sortAttrs = ['product'];
  } else if (listType === 'MusicList') {
    sortAttrs = ['artist', 'album', 'title'];
  } else if (listType === 'ToDoList') {
    sortAttrs = ['due_by', 'assignee_id', 'task'];
  }
  const sorted = performSort(items, sortAttrs);
  return sorted;
}

export async function fetchList({ id, history }) {
  try {
    // TODO: looks like list.users_list_id is not coming back
    const responses = await Promise.all([axios.get(`/lists/${id}/users_lists`), axios.get(`/lists/${id}`)]);
    const [
      {
        data: { accepted, pending },
      },
      {
        data: {
          current_user_id: currentUserId,
          not_purchased_items: responseNotPurchasedItems,
          purchased_items: purchasedItems,
          list,
          categories,
        },
      },
    ] = responses;
    const userInAccepted = accepted.find((acceptedList) => acceptedList.user.id === currentUserId);
    if (!userInAccepted) {
      toast('List not found', { type: 'error' });
      history.push('/lists');
      return;
    }
    const allAcceptedUsers = accepted.map(({ user }) => user);
    const allPendingUsers = pending.map(({ user }) => user);
    const listUsers = allAcceptedUsers.concat(allPendingUsers);
    const includedCategories = mapIncludedCategories(responseNotPurchasedItems);
    const notPurchasedItems = categorizeNotPurchasedItems(responseNotPurchasedItems, includedCategories);
    const { permissions } = userInAccepted.users_list;

    return {
      currentUserId,
      list,
      purchasedItems,
      categories,
      listUsers,
      includedCategories,
      notPurchasedItems,
      permissions,
    };
  } catch ({ response }) {
    if (response) {
      if (response.status === 401) {
        toast('You must sign in', { type: 'error' });
        history.push('/users/sign_in');
        return;
      } else if ([403, 404].includes(response.status)) {
        toast('List not found', { type: 'error' });
        history.push('/lists');
        return;
      }
    }
    // any other errors we will catch and render generic UnknownError
    throw new Error();
  }
}

export async function fetchListToEdit({ id, history }) {
  try {
    const {
      data: {
        list: { owner_id, id: listId, name, completed, type },
        current_user_id: currentUserId,
      },
    } = await axios.get(`/lists/${id}/edit`);
    if (owner_id !== currentUserId) {
      toast('List not found', { type: 'error' });
      history.push('/lists');
      return;
    }
    return {
      listId,
      name,
      completed,
      type,
    };
  } catch ({ response }) {
    if (response) {
      if (response.status === 401) {
        toast('You must sign in', { type: 'error' });
        history.push('/users/sign_in');
        return;
      } else if ([403, 404].includes(response.status)) {
        toast('List not found', { type: 'error' });
        history.push('/lists');
        return;
      }
    }
    // any other errors will just be caught and render the generic UnknownError
    throw new Error();
  }
}

export async function fetchItemToEdit({ itemId, listId, itemType, history }) {
  try {
    // TODO: do i need all of this?
    // TODO: confirm required items always come back
    const [editListResponse, usersListsResponse, listResponse] = await Promise.all([
      axios.get(`/lists/${listId}/${itemType}/${itemId}/edit`),
      axios.get(`/lists/${listId}/users_lists`),
      axios.get(`/lists/${listId}`),
    ]);
    const {
      data: {
        item,
        list: { id: returnedListId, type },
      },
    } = editListResponse;
    const {
      data: { accepted, pending, current_user_id: currentUserId },
    } = usersListsResponse;
    const {
      data: { categories },
    } = listResponse;
    const userId = item.user_id;
    const returnedItemId = item.id;
    const product = item.product || '';
    const task = item.task || '';
    const purchased = item.purchased || false;
    const quantity = item.quantity || '';
    const completed = item.completed || false;
    const author = item.author || '';
    const title = item.title || '';
    const read = item.read || false;
    const artist = item.artist || '';
    const album = item.album || '';
    const dueBy = formatDueBy(item.due_by); // TODO: this will default to today. ok?
    const acceptedUsers = accepted.map(({ user }) => user);
    const pendingUsers = pending.map(({ user }) => user);
    const listUsers = acceptedUsers.concat(pendingUsers);
    const userInAccepted = accepted.find((acceptedList) => acceptedList.user.id === currentUserId);
    const assigneeId = item.assignee_id ? String(item.assignee_id) : ''; // TODO: really?
    const numberInSeries = item.number_in_series ? Number(item.number_in_series) : 0;
    const category = item.category || '';
    if (!userInAccepted || userInAccepted.users_list.permissions !== 'write') {
      toast('Item not found', { type: 'error' });
      history.push(`/lists/${listId}`);
      return;
    }
    return {
      listUsers,
      userId,
      list: {
        id: returnedListId,
        type,
        categories,
      },
      item: {
        id: returnedItemId,
        product,
        task,
        purchased,
        quantity,
        completed,
        author,
        title,
        read,
        artist,
        dueBy,
        assigneeId,
        album,
        numberInSeries,
        category,
      },
    };
  } catch ({ response }) {
    if (response) {
      if (response.status === 401) {
        toast('You must sign in', { type: 'error' });
        history.push('/users/sign_in');
        return;
      } else if ([403, 404].includes(response.status)) {
        toast('Item not found', { type: 'error' });
        history.push(`/lists/${listId}`);
        return;
      }
    }
    // any other errors will just be caught and render the generic UnknownError
    throw new Error();
  }
}
