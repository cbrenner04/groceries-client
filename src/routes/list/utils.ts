import { toast } from 'react-toastify';
import { type AxiosError } from 'axios';

import axios from '../../utils/api';
import { EListType, type IList, type IListItem, type IListUser } from '../../typings';

export function itemName(item: IListItem, listType: EListType): string | undefined {
  return {
    BookList: `${item.title ? `"${item.title}"` : ''} ${item.author ?? ''}`,
    GroceryList: `${item.quantity ?? ''} ${item.product ?? ''}`,
    MusicList:
      `${item.title ? `"${item.title}"` : ''} ${item.artist ?? ''}` +
      `${item.artist && item.album ? ' - ' : ''}${item.album ?? ''}`,
    SimpleList: item.content,
    ToDoList: item.task,
  }[listType]?.trim();
}

export function mapIncludedCategories(items: IListItem[]): string[] {
  const cats = [''];
  items.forEach((item) => {
    if (!item.category) {
      return;
    }
    const cat = item.category.toLowerCase();
    if (!cats.includes(cat)) {
      cats.push(cat);
    }
  });
  return cats;
}

export function categorizeNotPurchasedItems(items: IListItem[], categories: string[]): Record<string, IListItem[]> {
  const obj: Record<string, IListItem[]> = {};
  categories.forEach((cat) => {
    obj[cat] = [];
  });
  items.forEach((item) => {
    if (!item.category) {
      obj[''].push(item);
      return;
    }
    const cat = item.category.toLowerCase();
    // TODO: why is this setting off this rule?
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    if (!obj[cat]) {
      obj[cat] = [];
    }
    obj[cat].push(item);
  });
  return obj;
}

function performSort(items: IListItem[], sortAttrs: (keyof IListItem)[]): IListItem[] {
  // return when all items are sorted
  if (sortAttrs.length === 0) {
    return items;
  }
  const sortAttr = sortAttrs.pop()!;
  const sorted = items.sort((a, b) => {
    const aSorter = a[sortAttr];
    const bSorter = b[sortAttr];
    // the sort from the server comes back with items with number_in_series: `null` at the end of the list
    // without the next two lines this would put those items at the front of the list
    if (!aSorter) {
      return 1;
    }
    if (!bSorter) {
      return -1;
    }
    const positiveBranch = aSorter > bSorter ? 1 : 0;
    return aSorter < bSorter ? -1 : positiveBranch;
  });
  return performSort(sorted, sortAttrs);
}

export function sortItems(listType: EListType, items: IListItem[]): IListItem[] {
  let sortAttrs: (keyof IListItem)[] = [];
  if (listType === EListType.BOOK_LIST) {
    sortAttrs = ['author', 'number_in_series', 'title'];
  } else if (listType === EListType.GROCERY_LIST) {
    sortAttrs = ['product'];
  } else if (listType === EListType.MUSIC_LIST) {
    sortAttrs = ['artist', 'album', 'title'];
  } else if (listType === EListType.SIMPLE_LIST) {
    sortAttrs = ['created_at', 'content'];
  } else {
    sortAttrs = ['due_by', 'assignee_id', 'task'];
  }
  const sorted = performSort(items, sortAttrs);
  return sorted;
}

function handleFailure(
  error: AxiosError,
  notFoundMessage: string,
  navigate: (url: string) => void,
  redirectURI: string,
): void {
  if (error.response) {
    if (error.response.status === 401) {
      toast('You must sign in', {
        type: 'error',
      });
      navigate('/users/sign_in');
      return;
    } else if ([403, 404].includes(error.response.status)) {
      toast(notFoundMessage, { type: 'error' });
      navigate(redirectURI);
      return;
    } else {
      toast(`Something went wrong. Data may be incomplete and user actions may not persist.`, { type: 'error' });
      return;
    }
  }
  // any other errors will just be caught and render the generic UnknownError
  throw new Error();
}

interface IFetchListReturn {
  currentUserId: string;
  list: IList;
  purchasedItems: IListItem[];
  categories: string[];
  listUsers: IListUser[];
  includedCategories: string[];
  notPurchasedItems: Record<string, IListItem[]>;
  permissions: string;
}

export async function fetchList(fetchParams: {
  id: string;
  navigate: (url: string) => void;
}): Promise<IFetchListReturn | undefined> {
  try {
    const { data } = await axios.get(`/lists/${fetchParams.id}`);
    const includedCategories = mapIncludedCategories(data.not_purchased_items);
    const notPurchasedItems = categorizeNotPurchasedItems(data.not_purchased_items, includedCategories);

    return {
      currentUserId: data.current_user_id,
      list: data.list,
      purchasedItems: data.purchased_items,
      categories: data.categories,
      listUsers: data.list_users,
      includedCategories,
      notPurchasedItems,
      permissions: data.permissions,
    };
  } catch (err: unknown) {
    handleFailure(err as AxiosError, 'List not found', fetchParams.navigate, '/lists');
  }
}

export async function fetchItemToEdit(fetchParams: {
  itemId: string;
  listId: string;
  navigate: (url: string) => void;
}): Promise<{ listUsers: IListUser[]; userId: string; list: IList; item: IListItem } | undefined> {
  try {
    const { data } = await axios.get(`/lists/${fetchParams.listId}/list_items/${fetchParams.itemId}/edit`);
    data.list.categories = data.categories;
    const userId = data.item.user_id;
    return {
      listUsers: data.list_users,
      userId,
      list: data.list,
      item: data.item,
    };
  } catch (err: unknown) {
    handleFailure(err as AxiosError, 'Item not found', fetchParams.navigate, `/lists/${fetchParams.listId}`);
  }
}

interface IFetchItemsToEditReturn {
  list: IList;
  lists: IList[];
  items: IListItem[];
  categories: string[];
  list_users: IListUser[];
}

export async function fetchItemsToEdit(fetchParams: {
  listId: string;
  search: string;
  navigate: (url: string) => void;
}): Promise<IFetchItemsToEditReturn | undefined> {
  try {
    const { data } = await axios.get(`/lists/${fetchParams.listId}/list_items/bulk_update${fetchParams.search}`);
    return data;
  } catch (err: unknown) {
    handleFailure(
      err as AxiosError,
      'One or more items not found',
      fetchParams.navigate,
      `/lists/${fetchParams.listId}`,
    );
  }
}
