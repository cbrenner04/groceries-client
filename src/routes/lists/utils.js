import { toast } from 'react-toastify';

import axios from '../../utils/api';

export const sortLists = (lists) => lists.sort((a, b) => new Date(b.created_at) - new Date(a.created_at));

function handleFailure({ response }, history) {
  // any other status code is super unlikely for these routes and will just be caught and render generic UnknownError
  if (response && response.status === 401) {
    toast('You must sign in', { type: 'error' });
    history.push('/users/sign_in');
    return;
  }
  throw new Error();
}

export async function fetchLists({ history }) {
  try {
    // TODO: confirm required items come back
    const { data } = await axios.get(`/lists/`);
    const userId = data.current_user_id;
    const sortedAcceptedLists = sortLists(data.accepted_lists);
    const pendingLists = sortLists(data.pending_lists);
    const completedLists = sortedAcceptedLists.filter((list) => list.completed);
    const nonCompletedLists = sortedAcceptedLists.filter((list) => !list.completed);
    const lists = sortedAcceptedLists.concat(pendingLists);
    const currentUserPermissions = {};
    // TODO: this information should be returned with the above request
    const usersListsRequests = await Promise.allSettled(
      lists.map((list) => axios.get(`/lists/${list.id}/users_lists/${list.users_list_id}`)),
    );
    const userLists = usersListsRequests
      .map((request) => {
        if (request.status === 'fulfilled') {
          return request.value;
        }
        return null;
      })
      .filter(Boolean);
    userLists.forEach((listData) => {
      const {
        data: { list_id: listId, permissions },
      } = listData;
      currentUserPermissions[listId] = permissions;
    });
    return {
      userId,
      pendingLists,
      completedLists,
      nonCompletedLists,
      currentUserPermissions,
    };
  } catch (error) {
    handleFailure(error, history);
  }
}

export async function fetchCompletedLists({ history }) {
  try {
    const {
      data: { completed_lists: completedLists },
    } = await axios.get(`/completed_lists/`);
    const currentUserPermissions = {};
    // TODO: this information should be returned with the above request
    const usersListsRequests = await Promise.allSettled(
      completedLists.map((list) => axios.get(`/lists/${list.id}/users_lists/${list.users_list_id}`)),
    );
    const userLists = usersListsRequests
      .map((request) => {
        if (request.status === 'fulfilled') {
          return request.value;
        }
        return null;
      })
      .filter(Boolean);
    userLists.forEach((listData) => {
      const {
        data: { list_id: listId, permissions },
      } = listData;
      currentUserPermissions[listId] = permissions;
    });
    return {
      completedLists,
      currentUserPermissions,
    };
  } catch (error) {
    handleFailure(error, history);
  }
}
