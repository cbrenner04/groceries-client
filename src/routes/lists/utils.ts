import { toast } from 'react-toastify';
import { type AxiosError } from 'axios';

import axios from 'utils/api';
import type { EListType, IList, TUserPermissions } from 'typings';

export const sortLists = (lists: IList[]): IList[] =>
  lists.sort((a, b) => Number(new Date(b.created_at!)) - Number(new Date(a.created_at!)));

function handleFailure(error: unknown, navigate: (url: string) => void): void {
  const err = error as AxiosError;
  // any other status code is super unlikely for these routes and will just be caught and render generic UnknownError
  if (err.response?.status === 401) {
    toast('You must sign in', { type: 'error' });
    navigate('/users/sign_in');
  } else {
    throw new Error(JSON.stringify(err));
  }
}

interface IFetchListsReturn {
  userId: string;
  completedLists: IList[];
  currentUserPermissions: TUserPermissions;
  pendingLists: IList[];
  incompleteLists: IList[];
}

export async function fetchLists(fetchParams: {
  navigate: (url: string) => void;
}): Promise<IFetchListsReturn | undefined> {
  try {
    const { data } = await axios.get('/lists/');
    const pendingLists = sortLists(data.pending_lists);
    const completedLists = sortLists(data.accepted_lists.completed_lists);
    const incompleteLists = sortLists(data.accepted_lists.not_completed_lists);
    return {
      userId: data.current_user_id,
      pendingLists,
      completedLists,
      incompleteLists,
      currentUserPermissions: data.current_list_permissions,
    };
  } catch (error) {
    handleFailure(error, fetchParams.navigate);
  }
}

export async function fetchCompletedLists(fetchParams: {
  navigate: (url: string) => void;
}): Promise<{ userId: string; completedLists: IList[]; currentUserPermissions: TUserPermissions } | undefined> {
  try {
    const { data } = await axios.get('/completed_lists/');
    return {
      userId: data.current_user_id,
      completedLists: data.completed_lists,
      currentUserPermissions: data.current_list_permissions,
    };
  } catch (error) {
    handleFailure(error, fetchParams.navigate);
  }
}

interface IFetchListToEditReturn {
  listId: string;
  name: string;
  completed: boolean;
  type: EListType;
}

export async function fetchListToEdit(fetchParams: {
  id: string;
  navigate: (url: string) => void;
}): Promise<IFetchListToEditReturn | undefined> {
  try {
    const { data } = await axios.get(`/lists/${fetchParams.id}/edit`);
    return {
      listId: data.id,
      name: data.name,
      completed: data.completed,
      type: data.type as EListType,
    };
  } catch (error: unknown) {
    const err = error as AxiosError;
    if (err.response) {
      if (err.response.status === 401) {
        toast('You must sign in', { type: 'error' });
        fetchParams.navigate('/users/sign_in');
      } else if ([403, 404].includes(err.response.status)) {
        toast('List not found', { type: 'error' });
        fetchParams.navigate('/lists');
      } else {
        // any other errors will just be caught and render the generic UnknownError
        throw new Error();
      }
    } else {
      // any other errors will just be caught and render the generic UnknownError
      throw new Error();
    }
  }
}

export function failure(error: unknown, navigate: (url: string) => void, setPending: (arg: boolean) => void): void {
  const err = error as AxiosError;
  if (err.response) {
    if (err.response.status === 401) {
      toast('You must sign in', { type: 'error' });
      navigate('/users/sign_in');
    } else if ([403, 404].includes(err.response.status)) {
      toast('List not found', { type: 'error' });
    } else {
      setPending(false);
      const responseTextKeys = Object.keys(err.response.data!);
      const responseErrors = responseTextKeys.map(
        (key) => `${key} ${(err.response?.data as Record<string, string>)[key]}`,
      );
      toast(responseErrors.join(' and '), { type: 'error' });
    }
  } else {
    setPending(false);
    const toastMessage = err.request ? 'Something went wrong' : err.message;
    toast(toastMessage, { type: 'error' });
  }
}

export function pluralize(listCount: number): string {
  return listCount > 1 ? 'Lists' : 'List';
}
