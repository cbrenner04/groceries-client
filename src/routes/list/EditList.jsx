import React from 'react';
import Async from 'react-async';
import PropTypes from 'prop-types';

import { fetchListToEdit } from './utils';
import EditListForm from './containers/EditListForm';
import Alert from '../../components/Alert';
import Loading from '../../components/Loading';

function EditList(props) {
  return (
    <Async promiseFn={fetchListToEdit} history={props.history} id={props.match.params.id}>
      <Async.Pending>
        <Loading />
      </Async.Pending>
      <Async.Fulfilled>
        {data => (
          <EditListForm
            history={props.history}
            listId={data.listId}
            name={data.name}
            completed={data.completed}
            type={data.type}
          />
        )}
      </Async.Fulfilled>
      <Async.Rejected>{error => <Alert errors={error.message} />}</Async.Rejected>
    </Async>
  );
}

EditList.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string,
    }),
  }),
  history: PropTypes.shape({
    push: PropTypes.func,
  }),
};

export default EditList;
