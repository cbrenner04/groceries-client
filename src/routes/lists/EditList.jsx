import React from 'react';
import Async from 'react-async';
import PropTypes from 'prop-types';

import { fetchListToEdit } from './utils';
import EditListForm from './containers/EditListForm';
import Loading from '../../components/Loading';
import UnknownError from '../error_pages/UnknownError';

function EditList(props) {
  return (
    <Async promiseFn={fetchListToEdit} history={props.history} id={props.match.params.id}>
      <Async.Pending>
        <Loading />
      </Async.Pending>
      <Async.Fulfilled>
        {(data) => (
          <EditListForm
            history={props.history}
            listId={data.listId}
            name={data.name}
            completed={data.completed}
            type={data.type}
          />
        )}
      </Async.Fulfilled>
      <Async.Rejected>
        <UnknownError />
      </Async.Rejected>
    </Async>
  );
}

EditList.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
};

export default EditList;
