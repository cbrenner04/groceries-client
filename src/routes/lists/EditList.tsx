import React from 'react';
import type { PromiseFn } from 'react-async';
import Async from 'react-async';
import { useNavigate, useParams } from 'react-router-dom';

import { fetchListToEdit } from './utils';
import EditListForm from './containers/EditListForm';
import Loading from '../../components/Loading';
import UnknownError from '../error_pages/UnknownError';

const EditList: React.FC = () => {
  const navigate = useNavigate();
  const { id } = useParams();

  return (
    // TODO: figure out typings for PromiseFn
    <Async promiseFn={fetchListToEdit as unknown as PromiseFn<void>} navigate={navigate} id={id}>
      <Async.Pending>
        <Loading />
      </Async.Pending>
      <Async.Fulfilled>
        {(data: { listId: string; name: string; type: string; completed: boolean }) => (
          <EditListForm listId={data.listId} name={data.name} completed={data.completed} type={data.type} />
        )}
      </Async.Fulfilled>
      <Async.Rejected>
        <UnknownError />
      </Async.Rejected>
    </Async>
  );
};

export default EditList;
