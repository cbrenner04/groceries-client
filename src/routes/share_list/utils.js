import axios from '../../utils/api';
import { setUserInfo } from '../../utils/auth';

export async function fetchData({ listId, history }) {
  try {
    const { data, headers } = await axios.get(`/lists/${listId}/users_lists`, {
      headers: JSON.parse(sessionStorage.getItem('user')),
    });
    setUserInfo(headers);
    const userInAccepted = data.accepted.find(acceptedList => acceptedList.user.id === data.current_user_id);
    if (!userInAccepted || !userInAccepted.users_list.permissions === 'write') {
      history.push('/lists');
      return;
    }
    return {
      name: data.list.name,
      invitableUsers: data.invitable_users,
      listId: data.list.id,
      userIsOwner: data.user_is_owner,
      pending: data.pending,
      accepted: data.accepted,
      refused: data.refused,
      userId: data.current_user_id,
    };
  } catch ({ response, request, message }) {
    const newError = new Error();
    if (response) {
      setUserInfo(response.headers);
      if (response.status === 401) {
        // TODO: how do we pass error messages along?
        history.push('/users/sign_in');
      } else if (response.status === 403) {
        // TODO: how do we pass error messages along
        history.push('/lists');
      } else {
        const responseTextKeys = Object.keys(response.data);
        const responseErrors = responseTextKeys.map(key => `${key} ${response.data[key]}`);
        newError.message = responseErrors.join(' and ');
      }
    } else if (request) {
      newError.message = 'Something went wrong!';
    } else {
      newError.message = message;
    }
    throw newError;
  }
}
