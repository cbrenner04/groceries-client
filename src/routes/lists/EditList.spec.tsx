import React from 'react';
import { render, type RenderResult, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router';

import EditList from './EditList';
import axios from '../../utils/api';
import * as utils from './utils';

jest.mock('react-router', () => ({
  ...jest.requireActual('react-router'),
  useNavigate: (): jest.Mock => jest.fn(),
  useParams: (): { id: string } => ({ id: '1' }),
}));

const setup = (): RenderResult =>
  render(
    <MemoryRouter>
      <EditList />
    </MemoryRouter>,
  );

describe('EditList', () => {
  it('renders the Loading component when fetch request is pending', async () => {
    const { container, findByText } = setup();

    expect(await findByText('Loading...')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('displays UnknownError when an error occurs', async () => {
    axios.get = jest.fn().mockRejectedValue({ message: 'failed to send request' });
    const { container, findByRole } = setup();
    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1));

    expect(await findByRole('button')).toHaveTextContent('refresh the page');
    expect(container).toMatchSnapshot();
  });

  it('displays UnknownError when data is undefined', async () => {
    // Mock the fetchListToEdit function to return undefined
    jest.spyOn(utils, 'fetchListToEdit').mockResolvedValue(undefined);
    const { container, findByRole } = setup();
    await waitFor(() => expect(utils.fetchListToEdit).toHaveBeenCalledTimes(1));

    expect(await findByRole('button')).toHaveTextContent('refresh the page');
    expect(container).toMatchSnapshot();
  });

  it('displays EditList', async () => {
    axios.get = jest.fn().mockResolvedValue({
      data: { owner_id: 'id1', id: 'id1', name: 'foo', completed: false, type: 'GroceryList' },
    });
    const { container, findByText } = setup();
    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1));

    expect(await findByText('Update List')).toBeVisible();
    expect(container).toMatchSnapshot();
  });
});
