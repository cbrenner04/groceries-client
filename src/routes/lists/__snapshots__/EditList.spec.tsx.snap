// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditList displays EditList 1`] = `
<div>
  <h1>
    Edit 
    foo
  </h1>
  <br />
  <form
    autocomplete="off"
    class=""
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="name"
      >
        Name
      </label>
      <input
        class="form-control"
        id="name"
        name="name"
        placeholder="My super cool list"
        type="text"
        value="foo"
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="type"
      >
        Type
      </label>
      <select
        class="form-control"
        id="type"
        name="type"
      >
        <option
          value="BookList"
        >
          books
        </option>
        <option
          value="GroceryList"
        >
          groceries
        </option>
        <option
          value="MusicList"
        >
          music
        </option>
        <option
          value="SimpleList"
        >
          simple
        </option>
        <option
          value="ToDoList"
        >
          to-do
        </option>
      </select>
    </div>
    <div
      class="mb-3 mb-3"
    >
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="completed"
          name="completed"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="completed"
          title=""
        >
          Completed
        </label>
      </div>
    </div>
    <div
      class="d-grid gap-2 mt-3"
    >
      <button
        class="btn btn-success"
        type="submit"
      >
        Update List
      </button>
      <button
        class="btn btn-link"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;

exports[`EditList displays UnknownError when an error occurs 1`] = `
<div>
  <h1>
    Something went wrong!
  </h1>
  <h2>
    We are currently unable to render this page.
  </h2>
  <p>
    Please check your connection and try to  
    <button
      style="background-color: transparent; cursor: pointer; display: inline; margin: 0px; padding: 0px; color: rgb(0, 123, 255); text-decoration: none;"
    >
      refresh the page
    </button>
  </p>
</div>
`;

exports[`EditList displays UnknownError when data is undefined 1`] = `
<div>
  <h1>
    Something went wrong!
  </h1>
  <h2>
    We are currently unable to render this page.
  </h2>
  <p>
    Please check your connection and try to  
    <button
      style="background-color: transparent; cursor: pointer; display: inline; margin: 0px; padding: 0px; color: rgb(0, 123, 255); text-decoration: none;"
    >
      refresh the page
    </button>
  </p>
</div>
`;

exports[`EditList renders the Loading component when fetch request is pending 1`] = `
<div>
  <div
    id="loader"
  >
    <div
      class="spinner-border"
      id="spinner"
      role="status"
    >
      <span
        class="visually-hidden"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
